### DISCUSSION TITLE
Child code-cells?

### DISCUSSION THREAD

**Discussion by mcanouil** [quarto-dev/quarto-cli#2094] [Q&A] [answered]
I tried different things, how to programmatically create child code cells and evaluate those child within Quarto? (My use case is a bit more complicated and related to #2088).

The following does not work for instance.
I can write a script that will produce a Quarto partial document to be included, but I'd rather avoid this two-step approach.

````
---
title: "Untitled"
format: html
keep-md: true
---

```{r}
#| output: asis
#| echo: false
cat(
  "```{r}",
  "#| eval: true",
  "#| echo: true",
  "1 + 1",
  "plot(1)",
  "```",
  sep = "\n"
)
```
````

It produces the following md file:

````md
---
title: "Untitled"
format: html
keep-md: true
---

```{r}
#| eval: true
#| echo: true
1 + 1
plot(1)
```
````

while I would have liked to have:

````md
---
title: "Untitled"
format: html
keep-md: true
---

::: {.cell}

```{.r .cell-code}
1 + 1
```

::: {.cell-output .cell-output-stdout}
```
[1] 2
```
:::

```{.r .cell-code}
plot(1)
```

::: {.cell-output-display}
![](test_files/figure-html/unnamed-chunk-2-1.png){width=672}
:::
:::
````

<details><summary>The actual use case Quarto document</summary>

````
---
title: "Untitled"
format: revealjs
---

```{r}
#| echo: false
#| output: asis
ggplot_code <- c(
  "library(palmerpenguins)",
  "library(ggplot2)",
  "",
  "ggplot(data = penguins) +", 
  "  aes(x = flipper_length_mm, y = body_mass_g) +",
  "  geom_point(aes(color = species, shape = species), size = 3, alpha = 0.8) +", 
  "  scale_color_manual(values = c(\"darkorange\", \"purple\", \"cyan4\")) +", 
  "  labs(",
  "    title = \"Penguin size, Palmer Station LTER\",", 
  "    subtitle = \"Flipper length and body mass for Adelie, Chinstrap and Gentoo Penguins\",", 
  "    x = \"Flipper length (mm)\",",
  "    y = \"Body mass (g)\",", 
  "    color = \"Penguin species\",",
  "    shape = \"Penguin species\"", 
  "  ) +",
  "  theme_minimal() +",
  "  theme(",
  "    legend.position = c(0.2, 0.7),", 
  "    plot.title.position = \"plot\",",
  "    plot.caption = element_text(hjust = 0, face = \"italic\"),", 
  "    plot.caption.position = \"plot\"",
  "  )"
)
offset <- 4
steps <- c(6, 7, 15, 16, 22)
for (i in steps) {
  cat(
    "\n## Slide 1\n",
    ":::: {.columns}",
    '::: {.column width="60%"}',
    "",
    "```{r}",
    "#| eval: false",
    "#| echo: true",
    paste0("#| code-line-numbers: 1-", i - offset + 1),
    paste(ggplot_code[offset:length(ggplot_code)], collapse = "\n"),
    "```",
    "",
    ":::",
    "",
    '::: {.column width="40%"}',
    "",
    "```{r}",
    "#| eval: true",
    "#| echo: false",
    "#| output: true",
    sub("\\+$", "", paste(ggplot_code[1:i], collapse = "\n")),
    "```",
    "",
    ":::",
    "::::",
    sep = "\n"
  )
}
```
````

</details>

**Comment by cderv** [comment, marked as answer]
What is exactly the aim of this ? Your full `.qmd` example seems a bit complicated to generated all this programmatically. 
Using an include file seems the way to go with Quarto. I think #2088 is more about issues in our supported feature, or regression from recent updates. Anyway, about your direct question see below. 

for R document, the **knitr** way can still be done. With R Markdown it can be possible using recipies from Cookbook which would work for Quarto as they are **knitr** supported. 
Those two recipies are about programmatically create content for the document leveraging 

* https://bookdown.org/yihui/rmarkdown-cookbook/child-document.html
* https://bookdown.org/yihui/rmarkdown-cookbook/knit-expand.html

Basically what is missing in your example is the computation engine processing (here **knitr** processing) that creates the results from your including code chunk inside code chunk. 

`````markdown
---
title: "Untitled"
format: html
keep-md: true
---

```{r}
#| echo: false
#| output: asis
child <- knitr::knit_child(text = c(
  "```{r}",
  "#| eval: true",
  "#| echo: true",
  "1 + 1",
  "```"), quiet = TRUE
)
cat(child, sep = "\n")
```
````

Usually, using `child` chunk option with a child document is a cleaner way to go. Include shortcode in Quarto should be the equivalent of that.

**Reply by mcanouil** [reply]
Oh, I forgot about that one function 🤦‍♂️

@cderv check out the rendered document in video to see how it is related to #2088 (secondary question)

https://user-images.githubusercontent.com/8896044/186428644-8c95bc87-4f89-46b4-8bdd-3e01d2422c20.mov


<details><summary>Quarto document</summary>

````
---
title: "Untitled"
format: revealjs
---

```{r}
#| output: asis
#| echo: false
ggplot_code <- c( # Better write the code in "code.r" and use "ggplot_code <- readlines('code.r')"
  "library(palmerpenguins)",
  "library(ggplot2)",
  "",
  "ggplot(data = penguins) +", 
  "  aes(x = flipper_length_mm, y = body_mass_g) +",
  "  geom_point(aes(color = species, shape = species), size = 3, alpha = 0.8) +", 
  "  scale_color_manual(values = c(\"darkorange\", \"purple\", \"cyan4\")) +", 
  "  labs(",
  "    title = \"Penguin size, Palmer Station LTER\",", 
  "    subtitle = \"Flipper length and body mass for Adelie, Chinstrap and Gentoo Penguins\",", 
  "    x = \"Flipper length (mm)\",",
  "    y = \"Body mass (g)\",", 
  "    color = \"Penguin species\",",
  "    shape = \"Penguin species\"", 
  "  ) +",
  "  theme_minimal() +",
  "  theme(",
  "    legend.position = c(0.2, 0.7),", 
  "    plot.title.position = \"plot\",",
  "    plot.caption = element_text(hjust = 0, face = \"italic\"),", 
  "    plot.caption.position = \"plot\"",
  "  )"
)
offset <- 4
steps <- c(6, 7, 15, 16, 22)
for (i in steps) {
  child <- knitr::knit_child(
    text = c(
      "\n## Slide 1\n",
      ":::: {.columns}",
      '::: {.column width="60%"}',
      "",
      "```{r}",
      "#| eval: false",
      "#| echo: true",
      paste0("#| code-line-numbers: 1-", i - offset + 1),
      paste(ggplot_code[offset:length(ggplot_code)], collapse = "\n"),
      "```",
      "",
      ":::",
      "",
      '::: {.column width="40%"}',
      "",
      "```{r}",
      "#| eval: true",
      "#| echo: false",
      "#| output: true",
      sub("\\+$", "", paste(ggplot_code[1:i], collapse = "\n")),
      "```",
      "",
      ":::",
      "::::"
    ), quiet = TRUE
  )
  cat(child, sep = "\n")
}
```

````
</details>

**Reply by cderv** [reply]
Ok. What I mean is that this should work without relying on such feature only available due to **knitr** support if `knitr_child` is used. 

Thanks for the workaround though ! Clever one showing the power of programming the content. 👍

**Reply by mcanouil** [reply]
Is there a way, to evaluate only the lines of code that are being highlighted?

Here, it's a bit more than a workaround, it allows to see the different figures while highlighting the code in a step-by-step manner.

**Reply by cderv** [reply]
Oh sorry I missed that !! That is cool ! I understand now why you need to problematically create the code chunks. 

In the R Markdown ecosystem, this would be feature you can found in this package: https://evamaerey.github.io/flipbookr/

I don't think there is yet an equivalent for Quarto but probably the 📦 above can be used with Quarto - I did not test it.

**Reply by mcanouil** [reply]
Thanks for the R package (I think I saw it in the past but did not look at it for some reasons).
I like the "Developed by The package maintainer." 🤣

**Reply by mcanouil** [reply]
It appears `flipbookr` is not compatible with Reveal.js and won't be (https://github.com/EvaMaeRey/flipbookr/issues/51).

**Reply by cderv** [reply]
A good project to add compatibility there ! Or in a quarto extensions 😉

**Reply by mcanouil** [reply]
Indeed, this is something really useful for workshop/tutorial/teaching, although it requires heavy static code analysis to split/evaluate properly all kind of code.

After a really quick search (I only looked at the first result): https://stackoverflow.com/questions/54987129/how-to-programmatically-create-several-new-cells-in-a-jupyter-notebook
So it is possible as well for Jupyter Notebook.

**Reply by cderv** [reply]
Good topic for a blog post 😉

**Reply by mcanouil** [reply]
Once I rebranded my website to Quarto 😅
I have materials to wrote posts about Quarto (for now I have a one post/year, I missed to make one on the 6th of May 2022 😱)
